### ISSUE TITLE
Differences in java compilation (javac)

### ISSUE THREAD

**Issue by jeffreytjs** [nus-cs2030/2021-s1#184]
# Description 
What's the difference between javac -d . *.java and javac *.java?

Is it the same because the former requires the inclusion of directory (which is . in this case) for compiling while the latter compiles within the current directory?

While solving project 1, I encountered this issue where the output generated locally were correct but it's not the case on codecrunch (screenshot attached below). I was compiling using javac *.java but I could not replicate the results from codecrunch even after redownloading the files to the same directory. 

Subsequently, following the advice by Wayne, I downloaded the files from codecrunch into a new directory and ran javac -d . *.java then I faced an issue where there is NoClassDefFoundError for cs2030.simulator.Server even though Server has been defined and the file is imported in Main. I don't have the full stacktrace now but if it's required I should be able to recreate it.

To be very honest, I am not sure if the issues are related. I don't quite get why there seem to be a cache of my files even after clearing all .class files and even recompiling.

## Topic: 
Project 1 

## Screenshots (if any): 
![photo6057859459440553075](https://user-images.githubusercontent.com/19368296/95705488-b5317380-0c86-11eb-88c1-2e5d285f5ada.jpg)

**Comment by lkldev** [comment]
Hi, you might have missed some class files in a different subdirectory. When we include the `package` statement into our code, java will attempt to find our file at the location specified by the `package` statement. So if your statement is `package cs2030.simulator.Simulator`, it will assume that the class file can be found at `cs2030/simulator/Simulator.class`. If you compile without the `-d` option (i.e. just `javac *.java`), it will proceed to compile all the classes in the current directory, ignoring where the files should be created(i.e. not packaged). With the `-d` option, it will take into account where the files should be created, and proceed to create these directories for you (i.e. packaged, with the files that need to be in `cs2030/simulator` created in that directory). Hope this helps :)

**Comment by jeffreytjs** [comment]
Oh yes, this is very helpful. I just now noticed that there are indeed the .class files in the subdirectory that I did not removed previously. I didn't know that it doesn't retrieve the .class files from my current dir and instead retrieved it from the name of the package. It has been great for my understanding.

One final quick check, so if I run `javac -d . *.java` again the .class files in cs2030/simulator will be overwritten accordingly right?
